File: packages/inputlayer-py/src/inputlayer/session.py
"""Session - ephemeral facts and rules (no + prefix)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from inputlayer.compiler import compile_bulk_insert, compile_insert, compile_rule
from inputlayer.relation import Relation

if TYPE_CHECKING:
    from inputlayer.connection import Connection
    from inputlayer.derived import Derived
    from inputlayer.result import ResultSet


class Session:
    """Manage session-scoped (ephemeral) data.

    Session inserts and rules omit the ``+`` prefix, making them ephemeral
    (cleared on disconnect or KG switch).
    """

    def __init__(self, connection: Connection) -> None:
        self._conn = connection

    async def insert(self, facts: Relation | list[Relation]) -> None:
        """Insert ephemeral session facts (no + prefix)."""
        if isinstance(facts, list):
            if not facts:
                return
            datalog = compile_bulk_insert(type(facts[0]), facts, persistent=False)
        else:
            datalog = compile_insert(facts, persistent=False)
        await self._conn.execute(datalog)

    async def define_rules(self, *targets: type[Derived]) -> None:
        """Define session-scoped rules (no + prefix)."""
        from inputlayer.derived import Derived

        for target in targets:
            head_name = Relation._resolve_name(target)
            head_columns = Relation._get_columns(target)
            for clause in target.rules:
                datalog = compile_rule(
                    head_name,
                    head_columns,
                    clause.select_map,
                    clause.relations,
                    clause.condition,
                    persistent=False,
                )
                await self._conn.execute(datalog)

    async def list_rules(self) -> list[str]:
        """List session rules."""
        result = await self._conn.execute(".session list")
        return [row[0] for row in result.rows] if result.rows else []

    async def drop_rule(
        self,
        name: str | None = None,
        *,
        index: int | None = None,
    ) -> None:
        """Drop a session rule by name, or a specific clause by index."""
        if name and index is not None:
            await self._conn.execute(f".session remove {name} {index}")
        elif name:
            await self._conn.execute(f".session drop {name}")
        else:
            raise ValueError("Must provide rule name")

    async def clear(self) -> None:
        """Clear all session facts and rules."""
        await self._conn.execute(".session clear")

File: packages/inputlayer-js/src/session.ts
/**
 * Session - ephemeral facts and rules (no + prefix).
 */

import type { Connection } from './connection.js';
import type { RelationDef } from './relation.js';
import type { Fact } from './types.js';
import type { RuleClause } from './compiler.js';
import { compileInsert, compileBulkInsert, compileRule } from './compiler.js';

/**
 * Manage session-scoped (ephemeral) data.
 *
 * Session inserts and rules omit the + prefix, making them ephemeral
 * (cleared on disconnect or KG switch).
 */
export class Session {
  private readonly conn: Connection;

  constructor(connection: Connection) {
    this.conn = connection;
  }

  /** Insert ephemeral session facts (no + prefix). */
  async insert(rel: RelationDef, facts: Fact | Fact[]): Promise<void> {
    const factList = Array.isArray(facts) ? facts : [facts];
    if (factList.length === 0) return;

    let iql: string;
    if (factList.length === 1) {
      iql = compileInsert(rel, factList[0], false);
    } else {
      iql = compileBulkInsert(rel, factList, false);
    }
    await this.conn.execute(iql);
  }

  /** Define session-scoped rules (no + prefix). */
  async defineRules(
    headName: string,
    headColumns: string[],
    clauses: RuleClause[],
  ): Promise<void> {
    for (const clause of clauses) {
      const iql = compileRule(headName, headColumns, clause, false);
      await this.conn.execute(iql);
    }
  }

  /** List session rules. */
  async listRules(): Promise<string[]> {
    const result = await this.conn.execute('.session list');
    return result.rows.length > 0 ? result.rows.map((row) => String(row[0])) : [];
  }

  /** Drop a session rule by name, or a specific clause by index. */
  async dropRule(name: string, index?: number): Promise<void> {
    if (index !== undefined) {
      await this.conn.execute(`.session remove ${name} ${index}`);
    } else {
      await this.conn.execute(`.session drop ${name}`);
    }
  }

  /** Clear all session facts and rules. */
  async clear(): Promise<void> {
    await this.conn.execute('.session clear');
  }
}
